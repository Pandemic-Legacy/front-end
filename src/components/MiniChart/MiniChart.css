.MiniChart {

  .container {
    text-align: left;
    display: inline-block;
    position: relative;
    width: 100%;
    /* Sets container ratio to 2:1 to match SVG ratio defined in D3 */
    padding-bottom: 50%; 
    vertical-align: top;
    overflow: visible;
  }

  svg {
    display: inline-block;
    position: absolute;
    top: 0;
    left: 0;
    overflow: visible;
  }

  h4 {
    margin-bottom: 0;
  }

}
