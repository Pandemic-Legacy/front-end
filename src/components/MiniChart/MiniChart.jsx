import React, { useRef, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { select, line, curveCardinal, axisBottom, axisRight, scaleLinear, scaleOrdinal, schemeCategory10, min, max } from 'd3';
import { useResizeObserver } from '../../hooks/d3Hooks';
import { Typography, makeStyles } from '@material-ui/core';
import styles from '../../styles/Chart.css';


export function MiniChart({ dataset, property }) {
  
  const svgRef = useRef();
  const wrapperRef = useRef();
  const dimensions = useResizeObserver(wrapperRef);
 
  function formatDate(badDate) {
    return badDate.toString().slice(6, 7) + '/' + badDate.toString().slice(8, 10);
  }  

  useEffect(() => {
    if(!dataset || !dataset.date) {
      console.log('No data, exiting useEffect()');
      return;
    }

    const svg = select(svgRef.current);

    // Setup variables and viewBox for responsive display
    // const { width, height } = dimensions || wrapperRef.current.getBoundingClientRect();
    // const { width } = dimensions || wrapperRef.current.getBoundingClientRect();
    // const height = width / 2;
    const width = 500;
    const height = 250;
    const margin = { top: 20, right: 40, bottom: 30, left: 0 };

    svg
      .attr('viewBox', `0, 0, ${width}, ${height}`)
      .attr('preserveAspectRatio', 'xMinYMin meet');

    // Define scales
    const xScale = scaleLinear()
      .domain([0, dataset['date'].length - 1]) // range of data
      .range([margin.left, width - margin.right]); // range of pixels
    const yScale = scaleLinear()
      .domain([-100, 100])
      .range([height - margin.bottom, margin.top]);
  
    // Define axis
    const xAxis = axisBottom(xScale)
      .ticks(dataset['date'].length / 10)
      .tickFormat(index => formatDate(dataset.date[index]));
    const yAxis = axisRight(yScale)
      .ticks(height / 40);
    

    // Draw axis on pre-existing elements
    svg
      .select(`.${styles.xAxis}`)
      .style('transform', `translateY(${height - margin.bottom}px)`)
      .call(xAxis);
    svg
      .select(`.${styles.yAxis}`)
      .style('transform', `translateX(${width - margin.right}px)`)
      .call(yAxis);

    // Draw background
    svg
      .append('rect')
      .attr('x', margin.left)
      .attr('y', margin.top)
      .attr('width', width - margin.left - margin.right)
      .attr('height', height - margin.top - margin.bottom)
      .attr('fill', 'none')
      .attr('class', styles.chartBackground);

    // Define line
    const myLine = line()
      .x((value, index) => xScale(index))
      .y(yScale)
      .curve(curveCardinal);

    // Draw line
    svg
      .selectAll('.graphLine')
      .data([dataset[property.key]])
      .join('path')
      .attr('class', 'graphLine')
      .attr('d', value => myLine(value))
      .attr('fill', 'none')
      .attr('stroke', 'red');

  }, [dataset]);

  return (   
    <div className={styles.Chart}>
      <Typography variant="h5">{property.description}</Typography>
      <div ref={wrapperRef} className={styles.container}>
        <svg ref={svgRef}>
          <g className={styles.xAxis} />
          <g className={styles.yAxis} />
        </svg>
      </div>
    </div>
  );
}

MiniChart.propTypes = {
  dataset: PropTypes.object.isRequired,
  property: PropTypes.object.isRequired
};


export default MiniChart;
