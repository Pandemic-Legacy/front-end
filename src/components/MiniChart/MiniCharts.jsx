import React, { useEffect } from 'react';
import { Grid } from '@material-ui/core';
import { useSelector, useDispatch } from 'react-redux';
import { getSelectedCountryCode, getMobilityChartData } from '../../selectors/selectors';
import { setMobilityChartDataByCountryCode } from '../../actions/actions';
import { MiniChart } from './MiniChart';
import { useStyles } from './MiniCharts.styles';


export const MiniCharts = () => {

  const selectedCountryCode = 'US'; // Placeholder for now
  const mobilityData = useSelector(getMobilityChartData);
  const dispatch = useDispatch();
  const styles = useStyles();


  useEffect(() => {
    dispatch(setMobilityChartDataByCountryCode(selectedCountryCode));
  }, [selectedCountryCode]);


  return (
    <>
      {console.log('in MiniCharts, mobilityData is', mobilityData)}
      <Grid container spacing={3} className={styles.root}>
        {/* Placeholders; Individualized charts incoming! */}
        <Grid item xs={6} sm={4}>
          <MiniChart dataSet={mobilityData} />
        </Grid>
        <Grid item xs={6} sm={4}>
          <MiniChart dataSet={mobilityData} />
        </Grid>
        <Grid item xs={6} sm={4}>
          <MiniChart dataSet={mobilityData} />
        </Grid>
        <Grid item xs={6} sm={4}>
          <MiniChart dataSet={mobilityData} />
        </Grid>
        <Grid item xs={6} sm={4}>
          <MiniChart dataSet={mobilityData} />
        </Grid>
        <Grid item xs={6} sm={4}>
          <MiniChart dataSet={mobilityData} />
        </Grid>
      </Grid>
    </>
  );
};

export default MiniCharts;
