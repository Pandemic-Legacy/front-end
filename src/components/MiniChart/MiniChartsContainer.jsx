import React, { useEffect } from 'react';
import { Grid } from '@material-ui/core';
import { useSelector, useDispatch } from 'react-redux';
import { getSelectedCountryCode, getMobilityChartData } from '../../selectors/selectors';
import { setMobilityChartDataByCountryCode } from '../../actions/actions';
import { MiniChart } from './MiniChart';


export const MiniChartsContainer = () => {

  const selectedCountryCode = useSelector(getSelectedCountryCode);
  const mobilityData = useSelector(getMobilityChartData);
  const dispatch = useDispatch();
  const properties = [
    { key: 'retailChange', description: 'Retail and recreation' },
    { key: 'groceryChange', description: 'Grocery and pharmacy' },
    { key: 'transitChange', description: 'Transit stations' },
    { key: 'parksChange', description: 'Parks' },
    { key: 'workplacesChange', description: 'Workplaces' },
    { key: 'residentialChange', description: 'Residential' }
  ];
    
  
  useEffect(() => {
    dispatch(setMobilityChartDataByCountryCode(selectedCountryCode));
  }, [selectedCountryCode]);


  return (
    <>
      <Grid container spacing={3}>
        { properties.map((property, i) => 
          <Grid item xs={6} sm={4} key={i}>
            <MiniChart dataset={mobilityData} property={property} />
          </Grid>
        )}
      </Grid>
    </>
  );
};

export default MiniChartsContainer;
