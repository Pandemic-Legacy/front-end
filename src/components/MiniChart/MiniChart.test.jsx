import React from 'react';
import { shallow } from 'enzyme';
import MiniChart from './MiniChart';
import { ThemeProvider } from '@material-ui/core';

describe('MiniChart component', () => {
  it('renders MiniChart', () => {
    const wrapper = shallow(
      <ThemeProvider>
        <MiniChart />
      </ThemeProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
