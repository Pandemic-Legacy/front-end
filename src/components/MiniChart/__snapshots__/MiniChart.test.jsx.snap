// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MiniChart component renders MiniChart 1`] = `
<ContextProvider>
  <MiniChart />
</ContextProvider>
`;
